google/cloud/aiplatform/training_jobs.py

```diff
@@ -15,9 +15,297 @@
+            name=self._ROOT_MODULE,
+            requirements=",".join(f'"{r}"' for r in self.requirements),
+            version=self._SETUP_PY_VERSION,
+        )
+
+        # Write setup.py
+        with setup_py_path.open("w") as fp:
+            fp.write(setup_py_output)
+
+        # Copy script as module of python package.
+        shutil.copy(self.script_path, script_out_path)
+
+        # Run setup.py to create the source distribution.
+        setup_cmd = self._SETUP_PY_SOURCE_DISTRIBUTION_CMD.format(
+            python_executable=_get_python_executable()
+        ).split()
+
+        p = subprocess.Popen(
+            setup_cmd,
+            cwd=trainer_root_path,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+        output, error = p.communicate()
+
+        # Raise informative error if packaging fails.
+        if p.returncode != 0:
+            raise RuntimeError(
+                "Packaging of training script failed with code %d\n%s \n%s"
+                % (p.returncode, output.decode(), error.decode())
+            )
+
+        return source_distribution_path
+
+    def package_and_copy(self, copy_method: Callable[[str], str]) -> str:
+        """Packages the script and executes copy with given copy_method.
+
+        Args:
+            copy_method Callable[[str], str]
+                Takes a string path, copies to a desired location, and returns the
+                output path location.
+        Returns:
+            output_path str: Location of copied package.
+        """
+
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            source_distribution_path = self.make_package(tmpdirname)
+            return copy_method(source_distribution_path)
+
+    def package_and_copy_to_gcs(
+        self,
+        staging_bucket: str,
+        project: str = None,
+        credentials: Optional[auth_credentials.Credentials] = None,
+    ) -> str:
+        """Packages script in Python package and copies package to GCS bucket.
+
+        Args
+            staging_bucket (str): Required. GCS Staging bucket. Should not contain gs://
+            project (str): Required. Project where GCS Staging bucket is located.
+            credentials (auth_credentials.Credentials):
+                Optional credentials used with GCS client.
+        Returns:
+            GCS location of Python package.
+        """
+
+        copy_method = functools.partial(
+            _timestamped_copy_to_gcs,
+            staging_bucket=staging_bucket,
+            project=project,
+            credentials=credentials,
+        )
+        return self.package_and_copy(copy_method=copy_method)
+
+    @property
+    def module_name(self) -> str:
+        """Module name that can be executed during training. ie. python -m"""
+        return f"{self._ROOT_MODULE}.{self._TASK_MODULE_NAME}"
+
+
+class TrainingJob(base.AiPlatformResourceNoun):
+
+    client_class = PipelineServiceClient
```

Comment:
Should be `JobServiceClient` instead?  Or perhaps the plan is to not expose CustomJob, requiring its access only through TrainingPipeline?

Reply:
The guidance we received was to use the PipelineService instead of the CustomJob service during review because it has stronger integration with the upstream and downstream resources. ie.  InputDataConfig takes a managed dataset, model_to_upload produces a managed model.